android/src/com/mapswithme/util/LocationUtils.java

```diff
@@ -98,6 +97,42 @@ private static boolean isSameLocationProvider(String p1, String p2)
     return (p1 != null && p1.equals(p2));
   }
 
+  public static boolean isFromGpsProvider(@NonNull Location location)
+  {
+    return LocationManager.GPS_PROVIDER.equals(location.getProvider());
+  }
+
+  public static boolean isFromFusedProvider(@NonNull Location location)
+  {
+    return LocationManager.FUSED_PROVIDER.equals(location.getProvider());
+  }
+
+  public static boolean isAccuracySatisfied(@NonNull Location location)
+  {
+    // If it's a gps location then we completely ignore an accuracy checking,
+    // because there are cases on some devices (https://jira.mail.ru/browse/MAPSME-3789)
+    // when location is good, but it doesn't contain an accuracy for some reasons
+    if (isFromGpsProvider(location))
+      return true;
+
+    // Completely ignore locations without lat and lon
+    return location.getAccuracy() > 0.0f;
+  }
+
+  public static boolean isLocationBetterThanLast(@NonNull Location newLocation, @NonNull Location lastLocation)
+  {
+    if (isFromFusedProvider(newLocation))
+      return true;
+
+    if (isFromGpsProvider(lastLocation) && lastLocation.getAccuracy() == 0.0f)
+      return true;
+
+    double speed = Math.max(DEFAULT_SPEED_MPS, (newLocation.getSpeed() + lastLocation.getSpeed()) / 2.0);
+    double lastAccuracy = lastLocation.getAccuracy() + speed * LocationUtils.getDiff(lastLocation, newLocation);
```

Comment:
nit: Rename getDiff with LocationUtils.getTimeDiff, because it also may be getLengthDiff, getSpeedDiff ...

Reply:
Done.